### PR TITLE
Implement SafeAreaEdges property and per-edge safe area control for Android

### PR DESCRIPTION
**Pull request by NirmalKumarYuvaraj**
> [!NOTE]
> Are you waiting for the changes in this PR to be merged?
> It would be very helpful if you could [test the resulting artifacts](https://github.com/dotnet/maui/wiki/Testing-PR-Builds) from this PR and let us know in a comment if this change resolves your issue. Thank you!

### Description of Change

This pull request introduces a unified and optimized approach for handling safe area insets and keyboard visibility across Android platform views in the codebase. The changes add a new `SafeAreaHandler` class, refactor `ContentViewGroup` and `LayoutViewGroup` to use this handler, and implement supporting types and extension methods for safe area calculations. The main goal is to ensure consistent, performant, and keyboard-aware safe area handling for layouts, reducing repeated hierarchy traversals and improving maintainability.

**Safe Area Handling Improvements**

* Added the new `SafeAreaHandler` class (`SafeAreaHandler.cs`) to encapsulate shared logic for safe area and keyboard inset handling, including caching, hierarchy traversal optimizations, and a unified window insets listener.
* Refactored `ContentViewGroup` and `LayoutViewGroup` to instantiate and use `SafeAreaHandler`, including setup of window insets listeners and safe area adjustment logic in their constructors and `OnLayout` methods. [[1]](diffhunk://#diff-5fdf872b32d4f8ee5971be75e1e5f45a53ec6c55fbdc701b54e02a175ca05ccaR17-R58) [[2]](diffhunk://#diff-3f4ba0ef39c71f2f98bc747387389cd17c63805a37849606234f7d8eb5d42e17R19-R65) [[3]](diffhunk://#diff-5fdf872b32d4f8ee5971be75e1e5f45a53ec6c55fbdc701b54e02a175ca05ccaR116-R121) [[4]](diffhunk://#diff-3f4ba0ef39c71f2f98bc747387389cd17c63805a37849606234f7d8eb5d42e17R129-R134)

**Supporting Types and Extensions**

* Introduced the `SafeAreaPadding` record struct and `WindowInsetsExtensions` static class for safe area and keyboard inset calculations, conversion, and rectangle adjustment.

**Dependency Updates**

* Added `AndroidX.Core.View` using statements to affected files to support new window insets APIs. [[1]](diffhunk://#diff-5fdf872b32d4f8ee5971be75e1e5f45a53ec6c55fbdc701b54e02a175ca05ccaR7) [[2]](diffhunk://#diff-3f4ba0ef39c71f2f98bc747387389cd17c63805a37849606234f7d8eb5d42e17R7)

These changes collectively provide a robust foundation for safe area and keyboard-aware layout on Android, making it easier to maintain and extend in the future.

### Issues Fixed

Fixes # https://devdiv.visualstudio.com/DevDiv/_workitems/edit/2433909

### Output

https://github.com/user-attachments/assets/4785f4c6-49ca-4f4e-9feb-519277b26942

### CODE REVIEW

**Comment by albyrock87** [comment]
Thank you so much for this effort!
I have some concerns regarding performance given this will be called on every node and it involves a lot of JNI calls.
Do you think we could move the majority of the logic on the Java side?

File: src/Core/src/Platform/Android/SafeAreaHandler.cs [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+            var parent = _owner.Parent;
+            int maxDepth = 10; // Limit traversal depth to prevent excessive hierarchy walking
+            int currentDepth = 0;
+
+            while (parent != null && currentDepth < maxDepth)
+            {
+                currentDepth++;
+
+                // Check if parent is a ContentViewGroup or LayoutViewGroup that responds to safe area
+                if (parent is ContentViewGroup parentContentGroup)
```
Comment (albyrock87) [inline src/Core/src/Platform/Android/SafeAreaHandler.cs]: What about having an interface implemented by both `ContentViewGroup` and `LayoutViewGroup` so that we can have this `Quick check` code only once? (even on the Java side of things)

File: src/Core/src/Platform/Android/ContentViewGroup.cs [hunk 2]
```diff
@@ -13,32 +14,48 @@ public class ContentViewGroup : PlatformContentViewGroup, ICrossPlatformLayoutBa
...
 		public ContentViewGroup(Context context, IAttributeSet attrs, int defStyleAttr, int defStyleRes) : base(context, attrs, defStyleAttr, defStyleRes)
 		{
 			_context = context;
+			_safeAreaHandler = new SafeAreaHandler(this, context, () => CrossPlatformLayout);
+			SetupWindowInsetsHandling();
+		}
+
+		void SetupWindowInsetsHandling()
+		{
+			ViewCompat.SetOnApplyWindowInsetsListener(this, _safeAreaHandler.GetWindowInsetsListener());
```
Comment (albyrock87) [inline src/Core/src/Platform/Android/ContentViewGroup.cs]: Pardon my ignorance, is it necessary to have a window inset listener on every single `View` or could we have a global listener?

File: src/Core/src/Platform/Android/SafeAreaHandler.cs (hunk 1 above)
Comment (albyrock87) [reply src/Core/src/Platform/Android/SafeAreaHandler.cs]: I think we already have one, https://github.com/dotnet/maui/blob/main/src/Core/src/Core/ICrossPlatformLayoutBacking.cs (but it's not on java side)

Comment (NirmalKumarYuvaraj) [reply src/Core/src/Platform/Android/SafeAreaHandler.cs]: @albyrock87 , I have updated the code changes. If you have any concerns please let me know

File: src/Core/src/Platform/Android/ContentViewGroup.cs (hunk 2 above)
Comment (albyrock87) [reply src/Core/src/Platform/Android/ContentViewGroup.cs]: @NirmalKumarYuvaraj so I've looked at some documentation and it appears to me that the window inset listener works as follows.

Every time the global window insets changes, Android fires the `onApplyWindowInsets()`on the root view, and then automatically propagates this even down to every view.

A given view can decide to consume the insets (aka do the safe area adjustment on itself) before returning them.

```
insets = insets.consumeSystemWindowInsets()
```

This way the descendants will receive reduced or zero insets.

That way we can avoid traversing the ancestors to detect whether the inset has been used or not.

I think we should probably leverage this feature if possible, what are your findings?

File: src/Core/src/Platform/Android/LayoutViewGroup.cs [hunk 3]
```diff
@@ -108,6 +126,12 @@ protected override void OnLayout(bool changed, int l, int t, int r, int b)
 
 			var destination = _context.ToCrossPlatformRectInReferenceFrame(l, t, r, b);
 
+			// Apply safe area adjustments if needed
+			if (_safeAreaHandler.RespondsToSafeArea())
```
Comment (albyrock87) [inline src/Core/src/Platform/Android/LayoutViewGroup.cs]: I've seen bad things happening on iOS due to measure pass not being in sync with layout pass.
I believe the measure pass should be aware of this too.

Comment (NirmalKumarYuvaraj) [reply src/Core/src/Platform/Android/LayoutViewGroup.cs]: @albyrock87 , I have updated the code changes. Please let me know if you have any concerns.

File: src/Core/src/Platform/Android/ContentViewGroup.cs (hunk 2 above)
Comment (NirmalKumarYuvaraj) [reply src/Core/src/Platform/Android/ContentViewGroup.cs]: @albyrock87 , The insets are still applied to every element in the view.

File: src/Core/src/Platform/Android/SafeAreaHandler.cs [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+        return oldSafeArea == _safeArea;
+    }
+
+#nullable disable
+    /// <summary>
+    /// WindowInsets listener for ContentViewGroup and LayoutViewGroup.
+    /// </summary>
+    public class WindowInsetsListener(SafeAreaHandler handler, Action requestLayout) : Java.Lang.Object, IOnApplyWindowInsetsListener
+    {
+        readonly SafeAreaHandler _handler = handler ?? throw new ArgumentNullException(nameof(handler));
```
Comment (jsuarezruiz) [inline src/Core/src/Platform/Android/SafeAreaHandler.cs]: The WindowInsetsListener holds strong references to SafeAreaHandler and Action. Couldould prevent garbage collection, can we update it to use a `WeakReference`?

`readonly WeakReference<SafeAreaHandler> _handlerRef;`

File: src/Core/src/Platform/Android/SafeAreaPadding.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+		return new Graphics.Rect(
+			bounds.X + Left,
+			bounds.Y + Top,
+			bounds.Width - HorizontalThickness,
+			bounds.Height - VerticalThickness);
+	}
+
+	public AInsets ToInsets() =>
+		AInsets.Of((int)Left, (int)Top, (int)Right, (int)Bottom) ?? AInsets.Of(0, 0, 0, 0)!;
```
Comment (jsuarezruiz) [inline src/Core/src/Platform/Android/SafeAreaPadding.cs]: `AInsets.Of()` should never return null for valid integers.

```
public AInsets ToInsets() =>
    AInsets.Of((int)Left, (int)Top, (int)Right, (int)Bottom);
```

**Review by jsuarezruiz — CHANGES_REQUESTED** [review]
These are the tests to validate the changes, right?
https://github.com/dotnet/maui/pull/30908/

Could we expand it to support more edge cases? I would include more tests like:

- `SafeAreaBorderOrientationPortraitToLandscape` - Tests orientation change from portrait to landscape, verifying that the safe area adapts, the border dimensions change, and insets adjust on notched devices.
- `SafeAreaBorderOrientationLandscapeToPortrait` - Tests orientation change from landscape to portrait, ensuring the safe area reverts correctly and border dimensions differ between orientations.
- `SafeAreaBorderMultipleOrientationChanges` - Tests multiple sequential orientation changes (portrait to landscape and back), verifying border visibility, positioning consistency within the same orientation (with tolerance), and overall stability.
- `SafeAreaBorderSoftInputBehavior` - Tests soft input (keyboard) behavior with bottom edge set to SoftInput, checking that the border shrinks when the keyboard appears and returns to original size after dismissal.
- `SafeAreaBorderSoftInputWithOrientationChange` - Tests soft input with orientation change (portrait to landscape) while keyboard is visible, ensuring the border maintains visibility and dimensions differ between orientations.

File: src/Core/src/Platform/Android/SafeAreaHandler.cs [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+    // Cached values for performance
+    ICrossPlatformLayout? _cachedLayout;
+    ISafeAreaView2? _cachedSafeAreaView2;
+    ISafeAreaView? _cachedSafeAreaView;
+    bool _layoutCacheValid;
+    internal IOnApplyWindowInsetsListener GetWindowInsetsListener() =>
+        new WindowInsetsListener(this, () => _owner.RequestLayout());
+    void InvalidateSafeArea()
+    {
+        _safeAreaInvalidated = true;
```
Comment (jsuarezruiz) [inline src/Core/src/Platform/Android/SafeAreaHandler.cs]: Could we protect here  to inconsistent state between `_safeAreaInvalidated` and `_layoutCacheValid` using locks?

**Review by jsuarezruiz — CHANGES_REQUESTED** [review]

File: src/Core/src/Platform/Android/SafeAreaHandler.cs (hunk 6 above)
Comment (albyrock87) [reply src/Core/src/Platform/Android/SafeAreaHandler.cs]: Is it really needed given UI changes must all run in the main thread? @jsuarezruiz

Comment (NirmalKumarYuvaraj) [reply src/Core/src/Platform/Android/SafeAreaHandler.cs]: @jsuarezruiz , Modified the code changes

File: src/Core/src/Platform/Android/SafeAreaPadding.cs (hunk 5 above)
Comment (NirmalKumarYuvaraj) [reply src/Core/src/Platform/Android/SafeAreaPadding.cs]: @jsuarezruiz , Removed the code. since it was not used anywhere

File: src/Core/src/Platform/Android/SafeAreaHandler.cs (hunk 4 above)
Comment (NirmalKumarYuvaraj) [reply src/Core/src/Platform/Android/SafeAreaHandler.cs]: @jsuarezruiz , Thanks for pointing this out. I have updated the code changes as requested.Please let me know if you have any concerns.

**Review by PureWeen — CHANGES_REQUESTED** [review]
I still need to review this a little bit deeper but I do still wonder about @albyrock87 's comment here

https://github.com/dotnet/maui/pull/31022#discussion_r2268725703

I wonder about the approach here with storing the safea area into separate handlers and then modifying the arrange pass. 

In theory could we just have a global handler that sets the padding on the view? 

Basically what I've done to fix the safearea on various views for edge to edge is just set the padding inside the windowinset listener

image

could we take the same approach here? Like, we just have a window listener applied to ContentViewGroup and LayoutViewGroup and then inside the "ApplyWindowLIstener" it extracts the safeareas and sets the padding on the views.

If the padding is set on the views than the measure and arrange passes will just work. The main reason we don't do this approach on iOS (AFAIK) is that iOS doesn't really have a concept of "padding" you can set on views, you always just fake the padding inside LayoutSubviews.

File: src/Core/src/Platform/Android/SafeAreaHandler.cs [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    /// WindowInsets listener for ContentViewGroup and LayoutViewGroup.
+    /// </summary>
+    public class WindowInsetsListener(SafeAreaHandler safeAreaHandler) : Java.Lang.Object, IOnApplyWindowInsetsListener
+    {
+        readonly WeakReference<SafeAreaHandler> _handlerRef = new(safeAreaHandler ?? throw new ArgumentNullException(nameof(safeAreaHandler)));
+
+        public WindowInsetsCompat OnApplyWindowInsets(View v, WindowInsetsCompat insets)
+        {
+            // Early return for null insets
+            if (insets is null)
```
Comment (albyrock87) [inline src/Core/src/Platform/Android/SafeAreaHandler.cs]: Let's improve this to `insets is null || !insets.hasInsets()`.

https://developer.android.com/reference/androidx/core/view/WindowInsetsCompat#hasInsets()

File: src/Core/src/PublicAPI/net-android/PublicAPI.Unshipped.txt [hunk 8]
```diff
@@ -58,6 +58,13 @@ Microsoft.Maui.LifecycleEvents.AndroidLifecycle.OnKeyLongPress
 Microsoft.Maui.LifecycleEvents.AndroidLifecycle.OnKeyMultiple
 Microsoft.Maui.LifecycleEvents.AndroidLifecycle.OnKeyShortcut
 Microsoft.Maui.LifecycleEvents.AndroidLifecycle.OnKeyUp
+Microsoft.Maui.Platform.IHandleWindowInsets
```
Comment (PureWeen) [inline src/Core/src/PublicAPI/net-android/PublicAPI.Unshipped.txt]: Let's keep these internal for now

File: src/Core/src/Platform/Android/MauiAppCompatActivity.cs [hunk 9]
```diff
@@ -9,6 +9,16 @@ namespace Microsoft.Maui
...
 	public partial class MauiAppCompatActivity : AppCompatActivity
 	{
+
+		private GlobalWindowInsetListener? _globalWindowInsetListener;
+
+		/// <summary>
+		/// Gets the shared GlobalWindowInsetListener instance for this activity.
+		/// This ensures all views use the same listener instance for coordinated inset management.
+		/// </summary>
+		public GlobalWindowInsetListener GlobalWindowInsetListener =>
```
Comment (PureWeen) [inline src/Core/src/Platform/Android/MauiAppCompatActivity.cs]: make this internal

File: src/Controls/src/Core/Platform/ModalNavigationManager/ModalNavigationManager.Android.cs [hunk 10]
```diff
@@ -311,8 +311,12 @@ public override AView OnCreateView(LayoutInflater inflater, ViewGroup? container
 
 				var rootView = _navigationRootManager?.RootView ??
 					throw new InvalidOperationException("Root view not initialized");
-				
-				ViewCompat.SetOnApplyWindowInsetsListener(rootView, new WindowHandler.WindowsListener());
+
+				var context = rootView.Context ?? inflater.Context;
+				if (context is not null)
+				{
+					rootView.SetGlobalWindowInsetListener(context);
```
Comment (PureWeen) [inline src/Controls/src/Core/Platform/ModalNavigationManager/ModalNavigationManager.Android.cs]: Should Modal pages use the same GlobalWindowInsetListener? It seems like the ModalNAvigaitonMAnager would have its own since it's going to be handling insets independently of the activity

File: src/Core/src/Platform/Android/LayoutViewGroup.cs [hunk 11]
```diff
@@ -21,28 +22,38 @@ public class LayoutViewGroup : ViewGroup, ICrossPlatformLayoutBacking, IVisualTr
...
 
 		public LayoutViewGroup(Context context, IAttributeSet attrs, int defStyleAttr, int defStyleRes) : base(context, attrs, defStyleAttr, defStyleRes)
 		{
 			_context = context;
+			SetupWindowInsetsHandling();
+		}
+
+		void SetupWindowInsetsHandling()
+		{
+			GlobalWindowInsetListenerExtensions.SetGlobalWindowInsetListener(this, _context);
```
Comment (PureWeen) [inline src/Core/src/Platform/Android/LayoutViewGroup.cs]: I wonder if this leaks........

I was going to say we need to make sure to set this to null but maybe it doesn't matter

We'll see how the device tests go but might be useful to see if this leaks.

What could work here instead is to assign this inside AttachedtoWindow and DeatchedFromWindow

For now let's just leave this but keep an eye on if this causes a leak...

I'm also curious if this fires once the view is detached from the activity....

Can you do a test here to see if you swap out the mainpage or remove the view if this still fires on views not part of the view hierarchy anymore?

File: src/Core/src/Platform/Android/MauiAppCompatActivity.cs (hunk 9 above)
Comment (NirmalKumarYuvaraj) [reply src/Core/src/Platform/Android/MauiAppCompatActivity.cs]: @PureWeen , Done, marked it as internal.

File: src/Core/src/PublicAPI/net-android/PublicAPI.Unshipped.txt (hunk 8 above)
Comment (NirmalKumarYuvaraj) [reply src/Core/src/PublicAPI/net-android/PublicAPI.Unshipped.txt]: Done, marked it as internal.

File: src/Core/src/Platform/Android/LayoutViewGroup.cs (hunk 11 above)
Comment (NirmalKumarYuvaraj) [reply src/Core/src/Platform/Android/LayoutViewGroup.cs]: @PureWeen , Yes , there was a leak with tracking , the tracked view count was increasing for each navigation. so created a new helper method to remove the listener.

Note : The listener is not called for the elements that are out of view.

File: src/Controls/src/Core/Platform/ModalNavigationManager/ModalNavigationManager.Android.cs (hunk 10 above)
Comment (NirmalKumarYuvaraj) [reply src/Controls/src/Core/Platform/ModalNavigationManager/ModalNavigationManager.Android.cs]: @PureWeen , Implemented separate listener for modal .

File: src/Core/src/Handlers/Window/WindowHandler.Android.cs [hunk 12]
```diff
@@ -16,10 +18,40 @@ public partial class WindowHandler : ElementHandler<IWindow, Activity>
 		protected override void ConnectHandler(Activity platformView)
 		{
 			base.ConnectHandler(platformView);
-
```
Comment (NirmalKumarYuvaraj) [inline src/Core/src/Handlers/Window/WindowHandler.Android.cs:19]: @PureWeen , FYI

| Before | After |
|--|--|
| Before | After |

File: src/TestUtils/src/UITest.Appium/HelperExtensions.cs [hunk 13]
```diff
@@ -2675,5 +2676,47 @@ public static bool IsElementVisible(IApp app, string elementName)
...
+		/// <returns>The display density as a double value (e.g., 1.0 for mdpi, 2.0 for xhdpi, 3.0 for xxhdpi)</returns>
+		public static double GetNativeDisplayDensity(this IApp app)
+		{
+			if (app is not AppiumApp aaa)
+			{
+				throw new InvalidOperationException($"GetNativeDisplayDensity is only supported on AppiumApp");
+			}
+
+			// For Android devices, use direct native capabilities
+			if (app is AppiumAndroidApp androidApp)
```
Comment (jsuarezruiz) [inline src/TestUtils/src/UITest.Appium/HelperExtensions.cs]: Could implement it in other platforms. For example, on iOS I think you can use`pixelRatio` or `scale` to get it.

Comment (jsuarezruiz) [reply src/TestUtils/src/UITest.Appium/HelperExtensions.cs]: Also can do something like:

```
var deviceScreenSize = (string?)((AppiumApp)App).Driver.Capabilities.GetCapability("deviceScreenSize")
	?? throw new InvalidOperationException("deviceScreenSize capability is missing or null.");
var deviceScreenDensity = (long?)((AppiumApp)App).Driver.Capabilities.GetCapability("deviceScreenDensity")
	?? throw new InvalidOperationException("deviceScreenDensity capability is missing or null.");
```

**Review by jsuarezruiz — CHANGES_REQUESTED** [review]
There are several testa failing, mostly all the Android tests using snapshots. The PR fundamentally changes how MAUI calculates and applies padding/margins for system UI elements. This is a breaking visual change that requires changes.
Currently we are cutting 60 top px on Android: https://github.com/dotnet/maui/blob/6d868559b690edcd60542b3c174c82f150fdab90/src/Controls/tests/TestCases.Shared.Tests/UITest.cs#L296


Example of a failing test, FontFamily:
`Snapshot different than baseline: LabelUITests_FontFamily_FontAwesome.png (17.88% difference)`
image

Next, the differences: 
image

File: src/TestUtils/src/UITest.Appium/HelperExtensions.cs (hunk 13 above)
Comment (NirmalKumarYuvaraj) [reply src/TestUtils/src/UITest.Appium/HelperExtensions.cs]: @jsuarezruiz , I have modified the code changes. please let me know if you have any concerns.

File: src/Controls/src/Core/SafeAreaElement.cs [hunk 14]
```diff
@@ -19,6 +22,22 @@ static void OnSafeAreaEdgesChanged(BindableObject bindable, object oldValue, obj
 			if (bindable is IView view)
 			{
 				view.InvalidateMeasure();
+
+#if ANDROID
+				// On Android, request layout does not call OnApplyWindowInsets. so we manually call it.
+				if (bindable is Element element && element.Handler is IElementHandler handler)
```
Comment (PureWeen) [inline src/Controls/src/Core/SafeAreaElement.cs]: I think you could achieve this with a mapper couldn't you? 

Add a mapper to MapSafeAreaEdges and have a specific one for android that calls this code?

The InvalidateMeasure should probably also be in a mapper

**Review by jsuarezruiz — DISMISSED** [review]
Build FAILED in .NET Standard:

```
/Users/builder/azdo/_work/1/s/src/Controls/src/Core/ContentPage/ContentPage.cs(183,25): error CS0117: 'OperatingSystem' does not contain a definition for 'IsIOS' [/Users/builder/azdo/_work/1/s/src/Controls/src/Core/Controls.Core.csproj::TargetFramework=netstandard2.1]
/Users/builder/azdo/_work/1/s/src/Controls/src/Core/ContentPage/ContentPage.cs(183,53): error CS0117: 'OperatingSystem' does not contain a definition for 'IsMacCatalyst' [/Users/builder/azdo/_work/1/s/src/Controls/src/Core/Controls.Core.csproj::TargetFramework=netstandard2.1]
/Users/builder/azdo/_work/1/s/src/Controls/src/Core/ContentPage/ContentPage.cs(183,25): error CS0117: 'OperatingSystem' does not contain a definition for 'IsIOS' [/Users/builder/azdo/_work/1/s/src/Controls/src/Core/Controls.Core.csproj::TargetFramework=netstandard2.0]
/Users/builder/azdo/_work/1/s/src/Controls/src/Core/ContentPage/ContentPage.cs(183,53): error CS0117: 'OperatingSystem' does not contain a definition for 'IsMacCatalyst' [/Users/builder/azdo/_work/1/s/src/Controls/src/Core/Controls.Core.csproj::TargetFramework=netstandard2.0]
    4 Error(s)

```

File: src/TestUtils/src/UITest.Appium/Actions/AppiumScrollActions.cs [hunk 15]
```diff
@@ -287,7 +287,7 @@ void ScrollToUp(AppiumDriver driver, AppiumElement element, ScrollStrategy strat
 
 			AppiumElement? result = ScrollUpUntilPresent(driver, target, element, strategy, swipePercentage, swipeSpeed, withInertia);
 
-			while (result is not null)
+			while (result is null)
```
Comment (PureWeen) [inline src/TestUtils/src/UITest.Appium/Actions/AppiumScrollActions.cs:290]: :-/

**Review by rmarinho — APPROVED** [review]
